app/src/org/koreader/launcher/AssetsUtils.java

```diff
@@ -0,0 +1,151 @@
+        try {
+            // is there any zip file inside the asset module?
+            String zipFile = getZipFile(context);
+            if (zipFile != null) {
+                // zipfile found! it will be extracted or not based on its version name
+                if (!isSameVersion(context, zipFile)) {
+                    InputStream stream = context.getAssets().open(zipFile);
+                    unzip(stream, output);
+                }
+                // extracted without errors.
+                return 1;
+            } else {
+                // check if the app has other, non-zipped, raw assets
+                Logger.i(TAG, "zip file not found, trying raw assets...");
+                return copyUncompressedAssets(context) ? 1 : 0;
+            }
+        } catch (IOException e) {
+            Logger.e(TAG, "error extracting assets:\n" + e.toString());
+            return 0;
+        }
+    }
+
+    /* do not extract assets if the same version is already installed */
+    private static boolean isSameVersion(Context context, String zipFile) {
+        try {
+            String output = context.getFilesDir().getAbsolutePath();
+            FileReader fileReader = new FileReader(output + "/git-rev");
+            BufferedReader bufferedReader = new BufferedReader(fileReader);
+            String installed_version = bufferedReader.readLine();
+            bufferedReader.close();
+            return version.contains(installed_version) ? true : false;
+        } catch (Exception e) {
+            Logger.d(TAG, "git revision not found, we should update");
+            return false;
+        }
+    }
+
+    /* get the first zip file inside the assets module */
+    private static String getZipFile(Context context) {
+        AssetManager assetManager = context.getAssets();
+        try {
+            String[] assets = assetManager.list("module");
+            for (String asset: assets) {
+                if (asset.endsWith(".zip")) {
+                    return "module/" + asset;
+                }
+            }
+            return null;
+        } catch (IOException e) {
+            Logger.e(TAG, "error listing assets:\n" + e.toString());
+            return null;
+        }
+    }
+
+    /* little programs don't need zipped assets.
+       ie: create a file under assets/module/ called llapp_main.lua,
+       and put there your start code. You can place other files under
+       assets/module and they will be copied too. */
+
+    private static boolean copyUncompressedAssets(Context context) {
+        AssetManager assetManager = context.getAssets();
+        String filesPath = context.getFilesDir().getAbsolutePath();
+        try {
+            // llapp_main.lua is the entry point for frontend code.
+            boolean entry_point;
+            String[] assets = assetManager.list("module");
+            for (String asset: assets) {
+                if ("llapp_main.lua".equals(asset) {
```

Comment:
You'll notice this soon enough, but:
```suggestion
                if ("llapp_main.lua".equals(asset)) {
```

Reply:
Err, actually the version you force pushed is a bit more broken. ;-)

Reply:
Ready now

Reply:
Cool! Trying to figure out how to refactor util due to the somewhat mysterious https://github.com/koreader/koreader/issues/5269 atm though. :-)

(I see where the circularity comes in, but not why I'm not getting it on desktop, Kobo or Android while this Kindle apparently is.)

Reply:
@pazos I can't figure out how this native libs bundling stuff works (or at least I can't in a minute or two and I don't want to waste my time on a solved problem).

I added this but it's ignoring the libs while creating the APK.
```diff
diff --git a/app/AndroidManifest.xml b/app/AndroidManifest.xml
index eb39428..aa84d79 100644
--- a/app/AndroidManifest.xml
+++ b/app/AndroidManifest.xml
@@ -19,6 +19,7 @@
         android:vmSafeMode="true"
         android:allowBackup="false"
         android:fullBackupContent="false"
+android:extractNativeLibs="true"
         android:usesCleartextTraffic="true">
         <meta-data android:name="android.max_aspect" android:value="3.1" />
         <activity
diff --git a/app/build.gradle b/app/build.gradle
index 77b2902..ce9c033 100644
--- a/app/build.gradle
+++ b/app/build.gradle
@@ -34,6 +34,7 @@ android {
             java.srcDirs = ['src']
             res.srcDirs = ['res']
             assets.srcDirs = [ '../assets' ]
+            jniLibs.srcDirs = [ '../libs' ]
         }
     }
```

Reply:
@Frenzie: now I just finished with this PR unless we catch some error. Shall I send another PR with the libraries branch? It should be easier for you to test (er, or at least we will share the same bugs :smile: )

Reply:
That sounds a lot easier, yes. If this is finished we can just go ahead and merge it I'd say, and go from there. :-)

Reply:
cool! we can merge this as-is, work on libraries and just update KOReader once we're done with both things.